aiida_sssp_workflow/workflows/convergence/engine.py

```diff
@@ -0,0 +1,156 @@
+            result_key=result_key,
+            convergence_window=convergence_window,
+            array_name=array_name,
+            current_index=current_index,
+            result_values=result_values,
+            initialized=initialized,
+            logger=logger,
+            result_state=result_state,
+        )
+        self.conv_thr = conv_thr
+
+    @property
+    def _num_new_iters(self) -> int:
+        """
+        Determine the minimum number of additional outputs to have a hope
+        of converging in the next step.
+        """
+        distance_triangle = self._distance_triangle
+        results_in_window = self._result_window
+        # Find location of the last calculation which creates out-of-tolerance
+        # roughness, and do enough calculations so that it is no longer in the
+        # next convergence window
+        num_new_iters_tol = 0
+        for i, row in enumerate(distance_triangle):
+            if np.any(np.array(row) > self.tol):
+                num_new_iters_tol = i + 1
+
+        # count the number of results that not satisfy conv_thr
+        # the more not satisfied the more add to the next iteration
+        # if all under the conv_thr add (zero) new iteration
+        num_new_iters_conv_thr = (np.array(results_in_window) >
+                                  self.conv_thr).sum()
+
+        # Use the max(greedy) number of the two types of criterion
+        num_new_iters = max(num_new_iters_tol, num_new_iters_conv_thr)
+
+        # Check that we don't go past the end of the input_values when trying
+        # to remove the calculation that is too rough from the window
+        # If we do, return -1 as an indication that convergence will not be
+        # possible
+        if self.current_index + num_new_iters > len(self.input_values):
+            # num_new_iters = len(self.input_values) - self.current_index
+            num_new_iters = -1
+
+        return num_new_iters
+
+    @property
+    def is_converged(self) -> bool:
+        """
+        Check if convergence has been reached by calculating the Frobenius
+        or 2-norm of the difference between all the result values / arrays
+        and checking that the maximum distance between points in the
+        convergence window is less than the tolerance.
+        """
+        if not self.initialized:
+            return False
+
+        # calculate pair distances between results
+        distance_triangle = self._distance_triangle
+        # flatten all the distances into a 1D list
+        distances = list(itertools.chain(*distance_triangle))
+
+        # check if the maximum distance is less than the tolerance
+        # import ipdb; ipdb.set_trace()
+        is_tol_converge = bool(np.max(distances) < self.tol)
+        is_conv_thr_converge = bool(
+            np.max(self._result_window) < self.conv_thr)
+
+        res = is_tol_converge and is_conv_thr_converge
+
+        return res
+
+
+class ConvergenceEngine(OptimizationEngineWrapper):
```

Comment:
I'd prefer to call this `TwoFactorConvergence`

Reply:
Ditto! Naming the class or the variable always the hardest thing ;-) especially for us non-native English speakers. I still remember the first time I was programming; I even use pinyin for the name of variables.